src/engine/renderer/tr_light.cpp

```diff
@@ -773,26 +770,28 @@ void R_SetupLightProjection( trRefLight_t *light )
 				VectorScale( right, ( 0.5f * dist ) / rLen, right );
 				VectorScale( up, - ( 0.5f * dist ) / uLen, up );
 
-				Vector4Set( lightProject[ 0 ], right[ 0 ], right[ 1 ], right[ 2 ], 0 );
-				Vector4Set( lightProject[ 1 ], up[ 0 ], up[ 1 ], up[ 2 ], 0 );
-				Vector4Set( lightProject[ 2 ], normal[ 0 ], normal[ 1 ], normal[ 2 ], 0 );
+				plane_t lightProject[ 4 ];
```

Comment:
`lightProject` doesn't look like planes. "PlaneMA" and "DotProductPlane" aren't operations that make sense for planes as far as I can tell.

Reply:
Maybe, but it is also processed separatly as veC3 and float, and feed some frustum:

```c++
VectorSubtract( lightProject[ 2 ], lightProject[ 1 ], frustum[ FRUSTUM_TOP ] );
frustum[ FRUSTUM_TOP ][ 3 ] = lightProject[ 2 ][ 3 ] - lightProject[ 1 ][ 3 ];
```

If it's not a plane, this would be the pefect candidate for a generic struct name for a `{ vec3, float }` couple which is actually processed like this.

One of the benefit of that change is that the day we use a better vector library that can do swizzles, it may be possible to do `something( lightProject.xyz, lightProject.t )` even if we go back to a vec4.

What I want to kill is to run vec3 things like `VectorSubstract` on a `vec4`.

Reply:
Maybe “it's not yet a plane but the result of the computation is a plane”?

Reply:
> Maybe, but it is also processed separatly as vec3 and float, and feeds some frustum:
> 
> ```c++
> VectorSubtract( lightProject[ 2 ], lightProject[ 1 ], frustum[ FRUSTUM_TOP ] );
> frustum[ FRUSTUM_TOP ][ 3 ] = lightProject[ 2 ][ 3 ] - lightProject[ 1 ][ 3 ];
> ```

This isn't actually treating the 4th component differently. It's just element-wise subtraction on vec4_t. Apparently VectorSubtract was used because there is no Vector4Subtract and the author wanted to code golf by doing it this way instead of writing out all 4 components or using a loop.

Reply:
Hmm, some operations may not be element-wise, for example:

```c++
VectorNegate( lightProject[ 3 ], frustum[ FRUSTUM_FAR ] );
frustum[ FRUSTUM_FAR ][ 3 ] = -lightProject[ 3 ][ 3 ] - 1.0f;
```
This is basically doing:

```c++
b[ 0 ] = -a[ 0 ];
b[ 1 ] = -a[ 1 ];
b[ 2 ] = -a[ 2 ];
b[ 3 ] = -a[ 3 ] - 1;
```

But this can be rewritten as:

```c++
Vector4Negate( a, b );
b[ 3 ] -= 1;
```

In the end this vec4 is copied to some frustum, so maybe I would need a Vec4ToPlane conversion.

Maybe w can consider this is some vector operation that builds-up a plane?

Reply:
I reverted this part to some `vec4_t` operation, adding `Vector4Subtract` and `Vector4Negate` macros, and adding a `PlaneSet( plane_t, vec4_t )` function to set the frustum planes using the computed vec4.

Reply:
> lightProject doesn't look like planes.

Those are, in fact, planes. Two of those are the right and up plane and one is near plane. The other planes are then derived using the near plane and the other 2 planes (since its normal is parallel to the lights direction apparently) and inverted (presumably to make the normals point outside of the frustum).

Plane dot product would presumably just be the dot product of their normals, i. e. represents the angle between planes.

The `ofs` that is calculated there seems to be used for shadowmaps (at least I don't think it can be non-parallel to the light target otherwise).

PlaneMA would then shift projection for the shadowmap.

See https://github.com/DaemonEngine/Daemon/blob/c5f8539fec2b29e7477c924e30c171d3e9fd2f65/src/engine/renderer/tr_main.cpp#L2327

(I haven't checked if DotProductPlane() and PlaneMA are correct to what old code was doing)

Reply:
> (I haven't checked if DotProductPlane() and PlaneMA are correct to what old code was doing)

This was assumed to be the same as Vector4MA, though maybe multiplying the 4th component was wrong if “Plane dot product would presumably just be the dot product of their normals”. When that light code was “working” (i.e. rendering things), the computes were broken anyway.

I would not be surprised if the plane dot product on 4 elements is wrong, since I expect to find bugs in that code.

@VReaperV I can revert my revert if you prefer.

Reply:
Well, we can merge this in current state (with things computed as vec4_t), where operations are done as vec4_t and then set to plane_t, with some comments for our future selves to fix and migrare that code later if needed.

Especially I would add a comment raising doubt about the dot product being done on 4 components.

Reply:
> Hmm, some operations may not be element-wise, for example:

That is the calculation for the far plane derived from near plane. The near and far planes are parallel, so the negation effectively inverses the normal (the normal of far and near plane are opposite). `start` is just the lights position, `falloff` is a normalized vector that determines how far the light effect goes I think. The `-DotProduct( start, falloff )` is then the distance to the near plane if the lights frustum was rotated around the origin (0; 0; 0) (negated to get the normal pointing outside of the frustum).

Note that VectorNegate does not negate the 4th component. The negation is done in `frustum[ FRUSTUM_FAR ][ 3 ] = -lightProject[ 3 ][ 3 ] - 1.0f` to set the the near and far planes to the [0; 1] range.

Reply:
> Note that VectorNegate does not negate the 4th component. 

Yes, that's exactly the kind of confusion I want to avoid with those patches.

My current code (using `vec4_t`) is doing:

```c++
Vector4Negate( lightProject[ 3 ], v );
v[ 3 ] -= 1.0f;
PlaneSet( frustum[ FRUSTUM_FAR ], v );
```

I added a `Vector4Negate` for this purpose.

Reply:
> This was assumed to be the same as Vector4MA, though maybe multiplying the 4th component was wrong if “Plane dot product would presumably just be the dot product of their normals”. When that light code was “working” (i.e. rendering things), the computes were broken anyway.

Well, considering we don't know (I assume) what exactly is breaking, we should keep the 4-component MA and dot, and maybe note them.

Reply:
Looks like it's doing the dot product of `(x y z 1)` and a plane to get the distance of `(x y z)` from the plane (with the result being positive on the "front" side or negative on the back). Maybe it would be better to use the 3-element dot product and then add the plane `dist` instead of building a vector with a dummy 4th value. That's how it is usually done in the tracing code for exmaple.

Reply:
Those 4th components were just adding 0 to 0 and mutiplying 0 with 0 anyway…

The first commit do the whole porting of the code of plane_t, but this functions still uses vec4_t as temporary variables before saving them to `plane_t` to keep the changes less intrusive.

The second commit deeply rewrites the `R_SetupLightProjection` function to never use `vec4_t` and to do the usual `VectorMA`, `DotProduct` on vec3. The `plane_t` is then set with this `vec3_t` normal and the related distance, which is zero in three cases out of four (hence the useless Vector4MA and DotProduct4).

The only plane not having distance set to zero was not using `Vector4MA` neither `DotProduct4`.

The now-useless `vec4_t`-related macros are now removed.